Add edge merging

zerebubuth opened the pull request:
This adds a function `valhalla::baldr::merge` which takes a `GraphReader` and a functor, and iterates over all the edges in the graph to find those which are collapsible, calling the functor for each collapsed set of edges.

This is a building block for OpenLR-like descriptors, and intended to group edges into sequences where there are no true intersections, but there might be bridges or changes in other properties such as max speed.

@kevinkreiser could you take a look and let me know what you think, please?

kevinkreiser commented on src/baldr/merge.cc:
```diff
@@ -0,0 +1,414 @@
+// 453 million ways in OSM, and we might allocate two edge IDs per way,
+// giving something like 108mb of bits needed.
+struct bitset_t {
+  bitset_t(size_t size) {bits.resize(size);}
```
there were some bugs with that bitset thing where it took up more memory than it needed: https://github.com/valhalla/tools/commit/d85fd7ab21087e03ed9470e0148bcd47d514b778#diff-5d0751e2f6630073e4f54144251387f0

zerebubuth replied:
Ooops, yeah. Not enough to have noticed on the small cases I've been testing on. Thanks for pointing it out, fixed in 0e1da23.

kevinkreiser commented on src/baldr/merge.cc:
```diff
@@ -0,0 +1,414 @@
+
+  //keep the global number of edges encountered at the point we encounter each tile
+  //this allows an edge to have a sequential global id and makes storing it very small
+  LOG_INFO("Enumerating edges...");
```
we shouldnt log in the lower level stuff we can just move logging to the calling application. same for log below

zerebubuth replied:
Yup, good call. Fixed in 6cda672.

kevinkreiser commented on src/baldr/merge.cc:
```diff
@@ -0,0 +1,414 @@
+    }
+  }
+
+  return std::move(edges_in_tiles);
```
is the explicit `move` necessary? i think the compiler is required to do the move for you in this case

zerebubuth replied:
Yup, that seems to be [the right way to do it](http://stackoverflow.com/questions/17473753/c11-return-value-optimization-or-move/17473874#17473874). I find the interaction between return value optimisation and move semantics to be quite confusing. Fixed in ffea013.

zerebubuth commented on src/baldr/merge.cc:
```diff
@@ -0,0 +1,414 @@
+        // this clears the cache, and the test relies on being able to inject
+        // stuff into the cache :-(
+        //
+        // TODO: presumably this is bad when the cache doesn't behave like a
```
Ah, I had missed the `GraphReader::OverCommitted()` function. Thanks for pointing it out. Fixed in 709ec6e.

kevinkreiser commented on src/baldr/merge.cc:
```diff
@@ -0,0 +1,432 @@
+
+namespace iter {
+
+// edges provides a container wrapper for the edges leaving a node which is
```
word `edges` at the beginning might have meant to be removed?

kevinkreiser commented on src/baldr/merge.cc:
```diff
@@ -0,0 +1,432 @@
+  // are the only two nodes reachable by non-shortcut edges, and none of the
+  // edges of @node_id cross into a different level.
+  boost::optional<std::pair<GraphId, GraphId> > nodes_reachable_from(GraphId node_id) {
+    boost::optional<GraphId> first, second;
```
instead of `optional<GraphId>` could we implement `GraphId::operator bool() const` and have it return `GraphId::Is_Valid()` or something?

kevinkreiser commented:
after boost optional comment above :rocket:

zerebubuth commented:
I removed optional in commit a8ac5ad, and added `GraphId`'s conversion to bool in c35edf7. That seems to build okay across the whole of Valhalla, so hopefully no unexpected changes in behaviour. It uses the new "implicit explicit" conversions, so should only happen implicitly for use in `if` statements and similar.

Added more flexibility in f923e32 and b360b75, which allow the customisation of the `GraphTile`s to iterate over and the edges to consider for collapsing, respectively. These can now be passed in as a range object and `std::function<>`, allowing the client code to decide the policy.

@kevinkreiser let me know what you think?
